ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -956,8 +959,23 @@ boolean getIndexEntriesSince(StoreKey key, FindEntriesCondition findEntriesCondi
    * @param entries the entries to eliminate duplicates from.
    */
   private void eliminateDuplicates(List<IndexEntry> entries) {
```

Comment:
I was looking at why you call `eliminateDuplicates()` on `entriesLocal` and not `entries` from within `getIndexEntriesSince()` (since for correctness, it would appear that eventually we want to eliminate duplicates for the whole list). It seems that the way `getEntriesSince()` is called, a new list is created and passed into it in all but one use in test. Even if that were not the case, I don't see a reason for `entriesLocal` to exist - that method could directly use the passed in list. Am I missing something?

Reply:
I am thinking if we can embed the deduplication in the for loop, because duplication happens only when set.size() > 1 and this case doesn't happen too often.

line 935:
```
      for (Map.Entry<StoreKey, ConcurrentSkipListSet<IndexValue>> entry : tempMap.entrySet()) {
        if (key == null || entry.getKey().compareTo(key) != 0) {
          if (! oneKeyPerEntry) {
               for (IndexValue value : entry.getValue()) {
                       IndexValue newValue = new IndexValue(startOffset.getName(), value.getBytes(), getVersion());
                       entriesLocal.add(new IndexEntry(entry.getKey(), newValue));
                       currentTotalSizeOfEntriesInBytes.addAndGet(value.getSize());
                }
          } else {
               lastIndexValueInTheSet = entry.getValue().last()
               if (isFlagSet(lastIndexValueInTheSet.Flags.Delete_Index)) {
                       // add this delete indexValue
                       IndexValue newValue = new IndexValue(startOffset.getName(), lastIndexValueInTheSet.getBytes(), getVersion());
                       entriesLocal.add(new IndexEntry(entry.getKey(), newValue));
                       currentTotalSizeOfEntriesInBytes.addAndGet(lastIndexValueInTheSet.getSize());
               } else {
                      firstIndexValueInTheSet = entry.getValue().first()
                      // add the first IndexValue(either PUT or TTL_UPDATE)
                      .... 
              }
          }

          // will break if size exceeded only after processing ALL entries for a key
          if (!findEntriesCondition.proceed(currentTotalSizeOfEntriesInBytes.get(), getLastModifiedTimeSecs())) {
            break;
          }
        }
      } 

```

Basically, the idea is to narrow the deduplication scope to set's size(3 max) instead of all.

Reply:
@vgkholla why does it have to run on the local list at all? Isn't it sufficient to run it once for the global list?

Reply:
Yeah there are use cases that need one entry per key from the `IndexSegment` - the dedup from here is required. The call from `PersistentIndex` _could_ skip this dedup (by setting the func param correctly) but I don't see it helping much - so I want to just keep the output from `IndexSegment::findEntriesSince()` like it was before.